deepspeed/runtime/engine.py

```diff
@@ -1763,10 +1764,10 @@ def allreduce_bucket(self, bucket, dp_group):
                     tensor_to_allreduce.mul_(self.gradient_predivide_factor() /
                                              dist.get_world_size(group=dp_group))
         else:
-            tensor_to_allreduce.div_(dist.get_world_size(group=dp_group))
+            tensor_to_allreduce.mul_(1. / dist.get_world_size(group=dp_group))
             dist.all_reduce(tensor_to_allreduce, group=dp_group)
 
-        if self.allreduce_always_fp32() and tensor is not tensor_to_allreduce:
+        if self.communication_data_type is not None and tensor is not tensor_to_allreduce:
```

Comment:
Do we want to copy if communication_data_type is fp16? Won't that deviate from previous behavior?

Reply:
Sorry, I was not very clear with my comment, so let me restate it. I think this condition should be 
`if self.condition_data_type == torch.float32`, so that it does not include when `condition_data_type == torch.float16`.